[TVOS] - make depends arm64 and tvos aware

Memphiz opened the pull request:
This is part 1 of the 3 major pull requests i plan to do for adding TVOS (AppleTV4) support to mainline. This basically bumps some depends and makes the depends addon system aware of the tvos platform.

For TVOS we need to compile for 64bit arm. That means that we get arm64 support for ios additionally as gift.

Part 2 will be arm64 support for core (we can compile ios 64bit builds with that)
Part 3 will be tvos support for core (needs a big amount of work before i even propose this as a PR).

I would like to get depends and arm64 support (meaning part 1 and 2) in soonish so that i don't need to carry that mass of commits in my tree.

As a side note - jenkins is already fully aware of arm64 ios builds and tvos builds.

Thx to @MrMC for the grunt work on tvos support and @linusyang for carrying arm64 support in is tree for what feels like ages :o).

Memphiz committed:
[depends] - add tvos support to depends configure and cleanup the darwin part a bit

Memphiz committed:
[depends/TexturePacker] - don't fail on already existing link

Memphiz committed:
[depends/boblight] - fix compilation for tvos (arm64/bitcode) and build and deploy libboblight by default for tvos

Memphiz committed:
[depends/target/config.site] - add arm64 support

Memphiz committed:
[depends/ffmpeg] - add compile support for ios/arm64

Memphiz committed:
[depends/fontconfig] - fix compilation for arm64

Memphiz committed:
[depends/gnutls/darwin] - fix compilation for tvos

Memphiz committed:
[depends/libffi] - version bump to 3.2.1 and fixed compilation for arm64

Memphiz committed:
[depends/libgcrypt] - fixed compilation for tvos

Memphiz committed:
[depends/libssh/ios] - fix compilation for tvos

Memphiz committed:
[depends/libvorbis] - fixed compilation for arm64

Memphiz committed:
[depends/mysql] - fixed arm64 compilation

Memphiz committed:
[depends/nettle/ios] - fixed compilation for arm64/bitcode

Memphiz committed:
[depends/openssl/ios] - link statically and don't build tests (fixes arm64/bitcode compilation)

Memphiz committed:
[depends/pcre] - version bump to 8.36 and fixed arm64/bitcode compilation

Memphiz committed:
[depends/python27/ios] - fixed compilation for arm64/bitcode and tvos (by disabeling fork, exec and system)

Memphiz committed:
[depends/pythonmodule-pil/pillow/ios] - fixed arm64 / bitcode compilation

Memphiz committed:
[depends/samba-gplv3] - fixed arm64/bitcode compilation (remove fork and exec which is not needed in libsmbclient)

Memphiz committed:
[depends/sqlite3] - don't build the sqlite3 binary program

Memphiz committed:
[depends] - libssh needs libgcrypt

Memphiz committed:
[depends] - define GASPP_FIX_XCODE5 for arm64 (needed by ffmpeg)

Memphiz committed:
[depends] - adapt readme with tvos support

Memphiz committed:
[depends] - updated gitignore

Memphiz committed:
[depends/curl] - fix tvos compilation (disable ntlm)

Memphiz commented:
jenkins build this please

Memphiz commented:
i think that one binary addon didn't compile on win32 due to a network timeout - unrelated to this PR...

wsnipex commented on tools/depends/README:
```diff
@@ -1,35 +1,49 @@
+    ./configure --host=arm-apple-darwin --with-cpu=arm64
+
+TVOS:
+    ./configure --host=arm-apple-darwin --with-cpu=arm64 --with-platform=tvos
```
isn't arm64 implicit on tvos?

Memphiz replied:
Not yet - but i will change configure.ac to make it implicit in case its not given on the configure line...

wsnipex commented on tools/depends/configure.ac:
```diff
@@ -9,6 +9,19 @@ m4_include([../../m4/xbmc_arch.m4])
+  AC_MSG_CHECKING([for $2])
+  case $1 in
+    $2*)
+      AC_MSG_ERROR(error in configure of --with-arch=$1)
```
--with-cpu

wsnipex commented on tools/depends/target/Makefile:
```diff
@@ -28,6 +28,9 @@ endif
 ifeq ($(OS),ios)
   EXCLUDED_DEPENDS = libcec libusb
+ifeq ($(CPU),arm64)
+  DEPENDS += boblight
```
???

Memphiz replied:
we build libboblight and bundle it - it needs to get signed during installation on tvos - i can add an check for tvos somehow i guess to make this not that missleading

wsnipex commented on tools/depends/target/fontconfig/fix-aarch64_atomics.patch:
```diff
@@ -0,0 +1,26 @@
+ #define fc_atomic_ptr_get(P)		(OSMemoryBarrier (), (void *) *(P))
+ #if (MAC_OS_X_VERSION_MIN_REQUIRED > MAC_OS_X_VERSION_10_4 || __IPHONE_VERSION_MIN_REQUIRED >= 20100)
++#if __aarch64__
++#define fc_atomic_ptr_cmpexch(P,O,N)	OSAtomicCompareAndSwap64Barrier ((int64_t) (O), (int64_t) (N), (int64_t*) (P))
```
cosmetics, this is the same as x86_64 and ppc64 below

Memphiz replied:
yes but ppc64 and x86_64 are in the else branch which we don't hit here

MrMC replied:
might be wrong but I think I pulled this patch from an upstream source.

wsnipex commented on tools/depends/target/libmicrohttpd/Makefile:
```diff
@@ -35,13 +35,13 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
-	$(MAKE) -C $(PLATFORM)/src/microhttpd
-	$(MAKE) -C $(PLATFORM)/src/include install
+	$(MAKE) -j1 -C $(PLATFORM)/src/microhttpd
+	$(MAKE) -j1 -C $(PLATFORM)/src/include install
```
why?

Memphiz replied:
because it fails when doing it multi threaded - at least it did in 4 of 5 cases ... i can limit it to darwin if you want

wsnipex replied:
it works fine on other platforms, so I wonder why it fails on tvos.

MrMC replied:
try a make -j20 on an i7 :)

Memphiz replied:
yeah might not be an issue on jenkins but had it with -j12 on an i7 - i'll try to figure out why

wsnipex commented on tools/depends/target/libxslt/Makefile:
```diff
@@ -29,7 +29,7 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
 
 $(LIBDYLIB): $(PLATFORM)
-	$(MAKE) -C $(PLATFORM)
+	$(MAKE) -j1 -C $(PLATFORM)
```
why?

wsnipex commented on tools/depends/target/python27/Makefile:
```diff
@@ -14,6 +14,7 @@ endif
 ifeq ($(OS),ios)
   HOSTPLATFORM=_PYTHON_HOST_PLATFORM="darwin"
   LINK_ICONV=-framework CoreFoundation -liconv
+  CFLAGS+=-DDISABLE_PROCESS_CREATION
```
if possible, we shouldn't continue our tradition of misusing CFLAGS for defines. This should be CPPFLAGS

Memphiz replied:
Since pythons configure doesn't pick it up and we provide it explicitley into the configure environment i can do so - but python is at fault anyway in this case ...

Memphiz replied:
doesn't fly - i think i simply change the patch to disable fork and co and apply it only for appletvos getting rid of the define ...

wsnipex replied:
yeah, cleaner anyway

wsnipex commented on tools/depends/target/samba-gplv3/Makefile:
```diff
@@ -42,6 +49,7 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
 ifeq ($(OS),android)
 	cd $(PLATFORM); patch -p0 < ../samba_android.patch
 endif
+	cd $(PLATFORM); patch -p0 < ../no_fork_and_exec.patch
```
please make this ios only

Memphiz commented:
Updated with the requested changes to configure, boblight, python, samba and i dropped the -j1 for libxslt and libmicrohttpd as it run through 2 times in a row (maybe the issue was with xcode7.1 i used when doing those commits). @davilla also have a look at the configure commit - i filled in target_platform for distinguishing iphoneos and appletvos (the variable was already there and used by rbpi) ...

MrMC commented:
looks fine

wsnipex commented:
one last question: is  tools/depends/target/libgcrypt/config.sub  there on purpose? Its 1,8k loc and not show here.

Memphiz commented:
good question - @davilla - whats this config.sub about? i removed it from the Makefile and it still builds fine here?

MrMC commented:
the config.sub in libgcrypt was stale, failed to build with original. maybe released to something else that was fixed in later commits. who knows. if it works without it, drop it and you can always re-add it if something shows up later.

MrMC commented:
I remember now.

checking build system type... Invalid configuration shared': machine shared not recognized configure: error: /bin/sh ./config.sub shared failed

my guess is https://github.com/Memphiz/xbmc/commit/1c6fbe70792e185ee69b290e3b4cc94704836b0e
fixed it.

Memphiz commented:
dropped the config.sub

Memphiz commented:
@wsnipex anything else? If not i would try jenkins for sanity again

wsnipex commented:
jenkins build this please

hudokkow commented:
Build is fine. pvr.hts keeps failing on Win32

Memphiz commented:
will merge tomorrow if no one else blocks it
